### PR TITLE
zm-mailbox RedoLog and RedoOp Test Coverage

### PR DESCRIPTION
**Pull request by spoon16**
We have increased the test coverage on RedoLog and RedoOp related classes. The goal of this work was to deeply familiarize ourselves with the RedoLog and RedoOp code and to better understand how to move forward developing a high availability mailbox worker pool.

- Increased test coverage across RedoLog and RedoOp related classes
- Changes to code to support assertions and development of above unit tests
- Removed some dead code
- Minor build cleanup
- Mark all broken tests as Ignore so that continuous integration will work
- Implement continuous integration using CircleCI

### CODE REVIEW

**Comment by fatkudu** [comment]
Please follow the links to sign the contributor license agreement.

File: common/src/java-test/com/zimbra/common/mime/MimeHeaderTest.java [hunk 1]
```diff
@@ -235,6 +236,7 @@ public void semiColonSeparated() {
 
     @SuppressWarnings("deprecation")
     @Test
+    @Ignore
```
Comment (fatkudu) [inline common/src/java-test/com/zimbra/common/mime/MimeHeaderTest.java]: Please raise a Jira ticket or Bugzilla ticket for tests that are being ignored (could group them by file perhaps)
and mention the ticket ID in the Ignore line.

File: store/src/java-test/com/zimbra/cs/redolog/op/CreateMessageTest.java [hunk 2]
```diff
@@ -12,17 +12,12 @@
 import com.zimbra.cs.mime.ParsedMessageDataSource;
 import com.zimbra.cs.redolog.RedoLogInput;
 import com.zimbra.cs.redolog.RedoLogOutput;
+import org.junit.*;
```
Comment (fatkudu) [inline store/src/java-test/com/zimbra/cs/redolog/op/CreateMessageTest.java]: Prefer to see individual imports.

**Comment by grishick** [comment]
LGTM. In future, I'd prefer if you broke changes down into smaller pull requests.

File: store/src/java-test/com/zimbra/cs/redolog/RedoLogManagerTest.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        redoLogManager.stop();
+    }
+
+    @Test
+    public void internalState() throws Exception {
+        Assert.assertEquals("build/test/redo/redo.log",
```
Comment (cgeisser) [inline store/src/java-test/com/zimbra/cs/redolog/RedoLogManagerTest.java]: Fixed up in as many places as I could find.

**Comment by gordyt** [comment]
## General Notes

Over all looks good.  Just want to point out a few things to look out for.

* Keep an eye out for mixed tabs/spaces in Java source.
* Try and make non-functional and Codacy commits first when updating an existing file.
    * Put all non-functional changes (whitespace, etc) in their own commit.  You can `git rebase -i` before submitting and fixup all of them into a single commit.
    * Put all codacy changes in their own commits.  Again, thay can be fixed-up into a single commit before submitting. The only reason for doing them separate from the non-functional commits is that a lot of the codacy (PMD) issues often involve a bit of code reorg so they are technically not-non-functional.
* In general, include Jira ticket numbers at the beginning of the first line of a commit message.  Having said that, as long as we are working out of separate Jira instances, that may be of limited value.
* Try to keep individual commits focused so there isn't a lot of non-related changes in the same commit.

## Ignored Unit Test Notes

Here is an overview of the `@Ignored` tests in the PR:

**Class** | **Ignored Tests**
--- | ----
AddHeaderTest | ALL
ContactAutoCompleteTest | ALL
ContactTest | ALL
DefangFilterTest | testBug88360, search
DeleteHeaderTest | ALL
EscapeSequencesTest | testReplaceheaderEscape1
EnvelopeTest | ALL
FlagTest | ALL
HeaderTest | ALL
MimeHeaderTest | semiColonSeparationDisabled
RejectTest | ALL
ReplaceHeaderTest | ALL
ServerThrottleTest | ALL
SetVariableTest | ALL


That seemed like a lot so I did five runs of the `zm-mailbox/store` unit tests (`ant clean compile test`) on the current `develop` branch and our `feature/imap` branch to see if I would see as many flaky test results as what you are seeing.  Results follow.

### develop branch

**Test** | **ERRRORS** | **FAILURES**
--- | ---- | ----
ContactAutoCompleteTest | 0 | 1
EnvelopeTest | 0 | 1
FlagTest | 1 | 0

**Test Run** | **ERRORS** | **FAIlURES**
--- | ---- | ----
1 | 0 | 0
2 | 0 | 0
3 | 0 | 1
4 | 0 | 1
5 | 1 | 0

### feature/imap branch

**Test** | **ERRRORS** | **FAILURES**
--- | ---- | ----
ContactAutoCompleteTest | 0 | 1
ContactTest | 2 | 1
FlagTest | 3 | 0
ImapHandlerTest | 0 | 2
MailboxTest | 0 | 1
RejectTest | 0 | 1

**Test Run** | **ERRORS** | **FAIlURES**
--- | ---- | ----
1 | 1 | 2
2 | 2 | 1
3 | 1 | 1
4 | 0 | 0
5 | 1 | 2

## Summary

I definitely have some flakyness in the unit tests as well, although no-where near as many. Note that I did not run the the `MimeHeaderTest` in `common`, just the tests in `store`. So there is more work to do on getting unit tests sorted out.  Since this PR is against the `develop` branch, I would like to get some feedback from the Platform team.  And Gren's comment regarding tickets for the newly-ignored tests is applicable.

**Comment by cgeisser** [comment]
Our test flakes come from running on circleci (visible here https://circleci.com/gh/f9teams/zm-mailbox ), which is a completely clean environment each time. I can't usually reproduce them locally, and some are probably configuration issues I haven't diagnosed yet.

**Comment by gordyt** [comment]
This PR (slightly modified [here](https://github.com/Zimbra/zm-mailbox/pull/270)) has been merged into the `develop` branch.

**Pull request closed without merging**